src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/SafeHandles.cs

```diff
@@ -238,4 +238,38 @@ protected sealed override bool ReleaseHandle()
             return true;
         }
     }
+
+    internal sealed class SafeChainEngineHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+        public SafeChainEngineHandle()
+            : base(true)
+        {
+        }
+
+        private SafeChainEngineHandle(IntPtr handle)
+            : base(true)
+        {
+            SetHandle(handle);
+        }
+
+        public static SafeChainEngineHandle MachineChainEngine =>
+            SafeHandleCache<SafeChainEngineHandle>.GetInvalidHandle(() => new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_LOCAL_MACHINE));
+
+        public static SafeChainEngineHandle UserChainEngine =>
+            SafeHandleCache<SafeChainEngineHandle>.GetInvalidHandle(() => new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_CURRENT_USER));
```

Comment:
These aren't modelling invalid handles, and only one of the two of them will be saved as "the" handle.

Instead they should just be

```C#
public static readonly SafeChainEngineHandle MachineChainEngine =
    new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_LOCAL_MACHINE);

public static readonly SafeChainEngineHandle UserChainEngine =
    new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_CURRENT_USER);
```

And Dispose:

```C#
protected override void Dispose(bool disposing)
{
    if (this != UserChainEngine && this != MachineChainEngine)
    {
        base.Dispose(disposing);
    }
}
```